### DISCUSSION TITLE
Inference

### DISCUSSION THREAD

**Discussion by npiroozan** [deepmodeling/deepmd-kit#1822] [Q&A]
Hello,

I hope you are well.  I've created a trained model based on Silicon DFT data and run inference with LAMMPS.  Results for RDF at 300K are consistent with DFT output for the same distribution function.  My next goal is to try and optimize inference throughput.

**Question:**

How can one output the Tensorboard trace for inference to see what the primary bottlenecks are and understand the forward propagation better?

Thank you very much!

**Comment by njzjz** [comment]
TensorFlow Profiler https://github.com/tensorflow/profiler

**Reply by npiroozan** [reply]
Thank for the information. I apologize for my naïveté regarding this.  How would one use this tool in DeePMD+LAMMPS?

Thank you very much for your time.

**Reply by njzjz** [reply]
I don't think there is any way to profile in C++ codes, but you can use the Python to do so with the same input.

**Reply by wanghan-iapcm** [reply]
The codes running on GPU, which are the most expensive parts of the deepmd+lammps, can be profiled by nvidia's profiling tools https://docs.nvidia.com/cuda/profiler-users-guide/index.html

**Reply by npiroozan** [reply]
Thank you very much for your help, I understand. I'll write a sample Python script for inference with input data and use the TensorFlow Profiler to trace the bottlenecks.

One additional follow on question if I may please:

I'm using the Silicon Data from the following paper:

Zhang, Linfeng, Jiequn Han, Han Wang, Wissam A. Saidi, Roberto Car, and Weinan E. "End-to-end Symmetry Preserving Inter-atomic Potential Energy Model for Finite and Extended Systems." arXiv preprint arXiv:1805.09003 (2018)

I'd like to plot the accuracy of my model with a Radial Distribution Function at 300K compared against that of the DFT data of Crystalline Silicon from the above paper.

How can one go about doing these computations with only box.npy, coord.npy, force.npy, and energy.npy available?

Warm Regards and thank you!

**Reply by wanghan-iapcm** [reply]
You are referred to a convenient but slow tool in dpdata https://github.com/deepmodeling/dpdata/blob/master/dpdata/md/rdf.py

**Reply by npiroozan** [reply]
Thank you very much!  You guys are the best.

**Reply by AnuragKr** [reply]
Hi @npiroozan ,

Hope you are doing well. Can you please explain steps how you have created rdf with only  box.npy, coord.npy, force.npy, and energy.npy available?

Warm regards and Thank you

**Reply by npiroozan** [reply]
Hi @AnuragKr,

Apologies for the late reply.

I hope you are doing well also.  Right, so if you use the code rdf.py, which is located in:

https://github.com/deepmodeling/dpdata/blob/master/dpdata/md/rdf.py

You can actually extract the rdf signature directly.  This will require a few steps first:

1.  sel_type = [0, None] - this refers to the number of atoms in your system.  Here, this means only one.
2. max_r = 6 - this refers to the maximum cutoff distance for the calculation of your rdf signature.
3. nbins=300 - this refers to the number of data points in your curve.  Be careful not to select too small or too high of a value.
4.  sys = dpdata.System('name of folder with data', fmt = 'deepmd/npy')

The last item above is arguably the most important.  Specify the data set you want to analyze as well as the type of data format.  For my case, numpy.  When everything is done properly, simply execute:

python rdf.py


I hope that helps.

**Reply by AnuragKr** [reply]
Hi,

Thanks, @npiroozan for explaining to me step-wise. I have encountered a few issues --

I am doing this with examples/water/data given in deepmd-kit repo.

1.  xx, stat = rdf(sys, sel_type = [0,1], max_r = 6, nbins = 100)
     Above line of code giving ValueError: too many values to unpack (expected 2)
  If we see the function it is returning 3 arrays xx,rdf,coord.I have considered the first 2 arrays for calculation, let me know whether it is correct or not.

2. If I have one system with multiple data folders should I calculate results in a loop and then combine them?

3. I want to generate result like this -- [link](https://github.com/CSIprinceton/workshop-july-2022/blob/main/hands-on-sessions/day-1/2-molecular-dynamics-deepmd/liquid_water/300K/H2O_RDF.ipynb). The result which we are getting following above code how can we calculate rdf for OO,OH,HH individually.

Warm regards and Thank you

@amcadmus  @njzjz  Please suggest if I am missing out on something.

**Reply by AnuragKr** [reply]
The above questions I solved.  
1. [Solution for 1 and 2 points](https://github.com/deepmodeling/dpdata/discussions/356)
2. Solution for 3 points -  We need to use the `sel_type` option of the RDF to calculate element-specific rdfs. if we have a type_map ['O', 'H'], then [0, 0] gives O-O RDF, [0, 1] gives O-H and [1,1] gives H-H.
